kv/kvserver/loqrecovery: TestRejectBadVersionApplication failed

kv/kvserver/loqrecovery.TestRejectBadVersionApplication [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11899228?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11899228?buildTab=artifacts#/) on master @ [b1a641f1d27653a4e93154926be57d0cae864bd6](https://github.com/cockroachdb/cockroach/commits/b1a641f1d27653a4e93154926be57d0cae864bd6):


```
=== RUN   TestRejectBadVersionApplication
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/95fc0dff57a5bf3a6a9cc9666d301bb7/logTestRejectBadVersionApplication4218276940
    test_log_scope.go:81: use -show-logs to present logs inline
    server_integration_test.go:664: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver/loqrecovery_test/pkg/kv/kvserver/loqrecovery/server_integration_test.go:664
        	Error:      	Should be true
        	Test:       	TestRejectBadVersionApplication
        	Messages:   	restarted node not found in verify status
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/95fc0dff57a5bf3a6a9cc9666d301bb7/logTestRejectBadVersionApplication4218276940
--- FAIL: TestRejectBadVersionApplication (8.39s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/replication
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRejectBadVersionApplication.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-31797

--- a/pkg/kv/kvserver/loqrecovery/server.go
+++ b/pkg/kv/kvserver/loqrecovery/server.go
@@ -663,7 +663,10 @@ func makeVisitAvailableNodes(
 				log.Infof(ctx, "visiting node n%d, attempt %d", node.NodeID, r.CurrentAttempt())
 				addr := node.AddressForLocality(loc)
 				var conn *grpc.ClientConn
-				conn, err = rpcCtx.GRPCDialNode(addr.String(), node.NodeID, rpc.DefaultClass).Connect(ctx)
+				// Note that we use ConnectNoBreaker here to avoid any race with probe
+				// running on current node and target node restarting. Errors from circuit
+				// breaker probes could confuse us and present node as unavailable.
+				conn, err = rpcCtx.GRPCDialNode(addr.String(), node.NodeID, rpc.DefaultClass).ConnectNoBreaker(ctx)
 				// Nodes would contain dead nodes that we don't need to visit. We can skip
 				// them and let caller handle incomplete info.
 				if err != nil {
@@ -745,7 +748,10 @@ func makeVisitNode(g *gossip.Gossip, loc roachpb.Locality, rpcCtx *rpc.Context) 
 			log.Infof(ctx, "visiting node n%d, attempt %d", node.NodeID, r.CurrentAttempt())
 			addr := node.AddressForLocality(loc)
 			var conn *grpc.ClientConn
-			conn, err = rpcCtx.GRPCDialNode(addr.String(), node.NodeID, rpc.DefaultClass).Connect(ctx)
+			// Note that we use ConnectNoBreaker here to avoid any race with probe
+			// running on current node and target node restarting. Errors from circuit
+			// breaker probes could confuse us and present node as unavailable.
+			conn, err = rpcCtx.GRPCDialNode(addr.String(), node.NodeID, rpc.DefaultClass).ConnectNoBreaker(ctx)
 			if err != nil {
 				if grpcutil.IsClosedConnection(err) {
 					log.Infof(ctx, "can't dial node n%d because connection is permanently closed: %s",

